File: server/src/main/java/net/bytemc/minestom/server/inventory/SingletonInventory.java
package net.bytemc.minestom.server.inventory;

import net.bytemc.minestom.server.inventory.item.Item;
import net.bytemc.minestom.server.inventory.item.impl.ClickableItem;
import net.kyori.adventure.text.Component;
import net.minestom.server.entity.Player;
import net.minestom.server.inventory.Inventory;
import net.minestom.server.inventory.InventoryType;
import net.minestom.server.item.ItemStack;
import org.jetbrains.annotations.NotNull;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class SingletonInventory implements InventoryConstruct {

    private final Inventory inventory;
    private final Map<Integer, Item> items = new ConcurrentHashMap<>();

    public SingletonInventory(InventoryType inventoryType, boolean clickable, String title) {
        this.inventory = new Inventory(inventoryType, title);
        this.inventory.addInventoryCondition((player, slot, clickType, result) -> {
            result.setCancel(!clickable);
            if (items.containsKey(slot) && items.get(slot) instanceof ClickableItem clickableItem) {
                clickableItem.click(player, clickType.name());
            }
        });
    }

    @Override
    public void setItem(int slot, @NotNull Item stack) {
        this.inventory.setItemStack(slot, stack.getItemStack());
        this.items.put(slot, stack);
    }

    @Override
    public void addItem(@NotNull Item stack) {
        var slot = this.getNextSlot();
        this.inventory.setItemStack(slot, stack.getItemStack());
        this.items.put(slot, stack);
    }

    @Override
    public void removeItem(int slot) {
        this.inventory.setItemStack(slot, ItemStack.AIR);
        this.items.remove(slot);
    }

    @Override
    public void removeItem(@NotNull Item clickableItem) {
        removeItem(this.getSlotByClickedItem(clickableItem));
    }

    @Override
    public void open(@NotNull Player player) {
        player.openInventory(this.inventory);
    }

    @Override
    public void clear() {
        this.inventory.clear();
    }

    @Override
    public void setTitle(String title) {
        this.inventory.setTitle(Component.text(title));
    }

    private int getNextSlot() {
        for (int i = 0; i < this.inventory.getSize(); i++) if (!items.containsKey(i)) return i;
        return -1;
    }

    private int getSlotByClickedItem(Item clickableItem) {
        for (Integer slot : this.items.keySet()) if (this.items.get(slot).equals(clickableItem)) return slot;
        return -1;
    }

    public Inventory getInventory() {
        return inventory;
    }

    public Map<Integer, Item> getItems() {
        return items;
    }
}

File: server/src/main/java/net/bytemc/minestom/server/inventory/SingletonInventory.kt
package net.bytemc.minestom.server.inventory

import net.bytemc.minestom.server.inventory.item.Item
import net.bytemc.minestom.server.inventory.item.impl.ClickableItem
import net.bytemc.minestom.server.inventory.item.impl.toggle.SwitchItem
import net.bytemc.minestom.server.inventory.item.impl.ToggleItem
import net.minestom.server.entity.Player
import net.minestom.server.inventory.Inventory
import net.minestom.server.inventory.InventoryType
import net.minestom.server.item.Material

open class SingletonInventory(var title: String, var type: InventoryType, var clickable: Boolean) {
    var inventory: Inventory = Inventory(type, title)
    private var items: HashMap<Int, Item> = HashMap()

    init {
        inventory.addInventoryCondition { player, slot, clickType, result ->
            val item = items[slot]
            if(item is ClickableItem) {
                item.click(player, clickType.name)
            } else if(item is SwitchItem) {
                item.click(player,this, slot)
            } else if(item is ToggleItem) {
                item.click(player,this, slot)
            }

            if (!clickable) {
                result.isCancel = true
            }
        }
    }

    fun open(player: Player) {
        player.openInventory(inventory)
    }

    fun fill(item: Item) {
        for(i in 0 until inventory.size) {
            if(inventory.itemStacks[i].material() == Material.AIR) {
                fill(i, item)
                break
            }
        }
    }

    fun fill(slot: Int, item: Item) {
        inventory.setItemStack(slot, item.itemStack)
        items[slot] = item
    }

    fun fill(row: Int, slot: Int, item: Item) {
        fill(((row - 1) * 9) + slot, item)
    }

    fun fillRow(row: Int, item: Item) {
        for(i in 0 until 9) {
            fill(row, i, item)
        }
    }
}